{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "060b4ae4-6141-4587-9040-7ea775b3c18a",
   "metadata": {},
   "source": [
    "Code for Ucell Scoring taken from https://github.com/BoevaLab/ANS_signature_scoring/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "79f4cbba-9849-43e7-9a84-452102209b93",
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Any, List, Optional, Tuple\n",
    "\n",
    "def check_signature_genes(var_names: List[str], gene_list: List[str], return_type: Any = list) -> List[str]:\n",
    "    \"\"\"\n",
    "    The method checks the availability of signature genes in the list of available genes (var_names). Genes not present in var_names are excluded.\n",
    "\n",
    "    Args:\n",
    "        var_names: List of available genes in the dataset.\n",
    "        gene_list: List of genes to score for (signature).\n",
    "        return_type: Indicates whether to return a list or a set.\n",
    "\n",
    "    Returns:\n",
    "        Filtered gene list.\n",
    "\n",
    "    Raises:\n",
    "        ValueError\n",
    "            if return_type is not of type list or set.\n",
    "            if gene_list gets empty.\n",
    "    \"\"\"\n",
    "    # assume that if single string is passed it represents a single gene that should be scored\n",
    "    if isinstance(gene_list, str):\n",
    "        gene_list = [gene_list]\n",
    "\n",
    "    if type(gene_list) not in [list, set]:\n",
    "        raise ValueError(f\"gene_list needs to be either list or set.\")\n",
    "\n",
    "    if len(gene_list) == 0:\n",
    "        raise ValueError(f'gene_list must contain at least 1 gene name in adata.var_names')\n",
    "\n",
    "    if len(gene_list) != len(set(gene_list)):\n",
    "        seen = set()\n",
    "        duplicates = [x for x in gene_list if x in seen or seen.add(x)]\n",
    "        warnings.warn(f'The passed gene_list contains duplicated genes: {duplicates}')\n",
    "\n",
    "    if return_type not in [list, set]:\n",
    "        raise ValueError(f\"return_type needs to be either list or set.\")\n",
    "\n",
    "    if not isinstance(var_names, list):\n",
    "        var_names = var_names.tolist()\n",
    "\n",
    "    genes_to_ignore = (set(gene_list)).difference(set(var_names))\n",
    "    gene_list = (set(var_names)).intersection(set(gene_list))\n",
    "\n",
    "    if len(genes_to_ignore) > 0:\n",
    "        sc.logging.warning(f\"genes are not in var_names and ignored: {sorted(list(genes_to_ignore))}\")\n",
    "    if len(gene_list) == 0:\n",
    "        raise ValueError(\"No valid genes were passed for scoring.\")\n",
    "\n",
    "    return return_type(gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d20b2cee-79df-45c5-9a0b-1f254b2e16bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "from typing import Optional, List\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from anndata import AnnData\n",
    "from joblib import Parallel, delayed\n",
    "from scanpy._utils import _check_use_raw\n",
    "from scipy.sparse import issparse, csr_matrix, isspmatrix_csr\n",
    "from scipy.stats import rankdata\n",
    "\n",
    "def u_stat(rank_value, maxRank=1500):\n",
    "    \"\"\"\n",
    "    The method computes the U statistic on signature gene ranks.\n",
    "\n",
    "    Args:\n",
    "        rank_value: Ranks of the signature genes.\n",
    "        maxRank: Cutoff for maximum rank allowed.\n",
    "\n",
    "    Returns:\n",
    "        The U statistic for given signature gene ranks.\n",
    "    \"\"\"\n",
    "    insig = rank_value > maxRank\n",
    "    if all(insig):\n",
    "        return 0\n",
    "    else:\n",
    "        rank_value[insig] = maxRank + 1\n",
    "        rank_sum = rank_value.sum()\n",
    "        len_sig = len(rank_value)\n",
    "        u_value = rank_sum - (len_sig * (len_sig + 1)) / 2\n",
    "        auc = 1 - u_value / (len_sig * maxRank)\n",
    "        return auc\n",
    "\n",
    "\n",
    "def compute_ranks_and_ustat(X_data, index, columns, gene_list, X_indices=None, X_indptr=None, X_shape=None,\n",
    "                            maxRank=1500):\n",
    "    \"\"\"\n",
    "    The following method computes for each cell in `X_data` the UCell score.\n",
    "\n",
    "    Args:\n",
    "        X_data: Current batch of gene expression data.\n",
    "        index: Index of cells.\n",
    "        columns: Names of genes.\n",
    "        gene_list: Signature genes.\n",
    "        X_indices: For sparse matrix reconstruction indices. If None, method assumes `X_data` to be a dense matrix.\n",
    "        X_indptr: For sparse matrix reconstruction index pointers. If None, method assumes `X_data` to be a dense matrix.\n",
    "        X_shape: For sparse matrix reconstruction shape of original matrix. If None, method assumes `X_data` to be\n",
    "            a dense matrix.\n",
    "        maxRank:  Cutoff for maximum rank allowed.\n",
    "\n",
    "    Returns:\n",
    "        For each cell in X_data the method returns the UCell score.\n",
    "    \"\"\"\n",
    "    if any([x is None for x in [X_indices, X_indptr, X_shape]]):\n",
    "        data_df = pd.DataFrame(\n",
    "            X_data, index=index, columns=columns\n",
    "        )\n",
    "    else:\n",
    "        data_df = pd.DataFrame(\n",
    "            csr_matrix((X_data, X_indices, X_indptr), X_shape, copy=False).todense(), index=index, columns=columns\n",
    "        )\n",
    "\n",
    "    res = (data_df.apply(\n",
    "        lambda x: rankdata(-x),\n",
    "        axis=1,\n",
    "        raw=True\n",
    "    ))[gene_list]\n",
    "\n",
    "    del data_df\n",
    "\n",
    "    score = res.apply(\n",
    "        func=(lambda x: u_stat(x, maxRank=maxRank)),\n",
    "        axis=1,\n",
    "        raw=False\n",
    "    )\n",
    "    return score\n",
    "\n",
    "\n",
    "def score_genes(\n",
    "        adata: AnnData,\n",
    "        gene_list: List[str],\n",
    "        maxRank: int = 1500,\n",
    "        bs: int = 500,\n",
    "        score_name: str = \"UCell_score\",\n",
    "        random_state: Optional[int] = None,\n",
    "        copy: bool = True,\n",
    "        use_raw: Optional[bool] = None,\n",
    "        verbose: int = 0,\n",
    "        joblib_kwargs: dict = {'n_jobs': 4}\n",
    ") -> Optional[AnnData]:\n",
    "    \"\"\"\n",
    "\n",
    "    UCell signature scoring method is a Python implementation of the scoring method proposed by Andreatta et al. 2021.\n",
    "\n",
    "    Massimo Andreatta and Santiago J Carmona. „UCell: Robust and scalable single-cell\n",
    "    gene signature scoring“. en. In: Comput. Struct. Biotechnol. J. 19 (June 2021),\n",
    "    pp. 3796–3798 (cit. on pp. iii, 2, 9, 15, 16).\n",
    "\n",
    "    Implementation is inspired by score_genes method of Scanpy\n",
    "    (https://scanpy.readthedocs.io/en/latest/generated/scanpy.tl.score_genes.html#scanpy.tl.score_genes)\n",
    "\n",
    "    Args:\n",
    "        adata: AnnData object containing the gene expression data.\n",
    "        gene_list: A list of genes (signature) for which the cells are scored for.\n",
    "        maxRank: Cutoff for maximum rank allowed.\n",
    "        bs: The number of cells in a processing batch.\n",
    "        score_name: Column name for scores added in `.obs` of data.\n",
    "        random_state: Seed for random state.\n",
    "        copy: Indicates whether original or a copy of `adata` is modified.\n",
    "        use_raw: Whether to compute gene signature score on raw data stored in `.raw` attribute of `adata`\n",
    "        verbose: If verbose is larger than 0, print statements are shown.\n",
    "        joblib_kwargs: Keyword argument for parallel execution with joblib.\n",
    "\n",
    "    Returns:\n",
    "        If copy=True, the method returns a copy of the original data with stored UCell scores in `.obs`, otherwise\n",
    "        None is returned.\n",
    "    \"\"\"\n",
    "    start = sc.logging.info(f\"computing score {score_name!r}\")\n",
    "    if verbose > 0:\n",
    "        print(f\"computing score {score_name!r}\")\n",
    "\n",
    "    adata = adata.copy() if copy else adata\n",
    "\n",
    "    use_raw = _check_use_raw(adata, use_raw)\n",
    "\n",
    "    _adata = adata.raw if use_raw else adata\n",
    "\n",
    "    if random_state is not None:\n",
    "        np.random.seed(random_state)\n",
    "\n",
    "    # remove genes from gene_list not available in the data\n",
    "    gene_list = check_signature_genes(_adata.var_names, gene_list)\n",
    "\n",
    "    # check type of rank\n",
    "    if not isinstance(maxRank, int):\n",
    "        raise ValueError(f\"maxRank {maxRank} must be of type int\")\n",
    "\n",
    "    # check maxRank is not larger than available nr. of genes\n",
    "    if maxRank > len(_adata.var_names):\n",
    "        print(\n",
    "            f\"Provided maxRank is larger than the number of available genes. Set maxRank=len(adata.var_names)\"\n",
    "        )\n",
    "        maxRank = len(_adata.var_names)\n",
    "\n",
    "    # check that signature is not longer than maxRank\n",
    "    if maxRank < len(gene_list) <= len(_adata.var_names):\n",
    "        warnings.warn(\n",
    "            f\"The provided signature contains more genes than the maxRank parameter. maxRank is increased to \"\n",
    "            f\"signature length \"\n",
    "        )\n",
    "        maxRank = len(gene_list)\n",
    "\n",
    "    sparse_X = issparse(_adata.X)\n",
    "    if sparse_X and not isspmatrix_csr(adata.X):\n",
    "        adata.X = adata.X.tocsr()\n",
    "        warnings.warn(\"Chaning adata.X format to CSR format\")\n",
    "        # create groups of managable sizes\n",
    "    bss = pd.cut(np.arange(_adata.obs.shape[0]), (_adata.obs.shape[0] // bs + 1), labels=False)\n",
    "\n",
    "    scores = Parallel(**joblib_kwargs)(\n",
    "        delayed(compute_ranks_and_ustat)(\n",
    "            X_data=_adata[group[1].index,].X.data if sparse_X else _adata[group[1].index,].X,\n",
    "            X_indices=_adata[group[1].index,].X.indices if sparse_X else None,\n",
    "            X_indptr=_adata[group[1].index,].X.indptr if sparse_X else None,\n",
    "            X_shape=_adata[group[1].index,].X.shape if sparse_X else None,\n",
    "            index=group[1].index,\n",
    "            columns=_adata.var_names,\n",
    "            gene_list=gene_list,\n",
    "            maxRank=maxRank) for group in _adata.obs.groupby(bss))\n",
    "    scores = pd.concat(scores)\n",
    "\n",
    "    adata.obs[score_name] = scores\n",
    "\n",
    "    sc.logging.info(\n",
    "        \"    finished\",\n",
    "        time=start,\n",
    "        deep=(\"added\\n\" f\"    {score_name!r}, score of gene set (adata.obs).\"),\n",
    "    )\n",
    "    return adata if copy else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "8e47c1f4-3079-453e-9e5b-49315bac5c7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 10856 × 32146\n",
       "    obs: 'batch', 'tissue', 'doublet', 'doublet_score'\n",
       "    var: 'gene_ids', 'n_cells'"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad('A01_RAW.h5ad')\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "8063d18f-612a-46da-a782-00198707b44b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['OR4F5', 'RP11-34P13.7', 'RP11-34P13.8', 'RP11-34P13.9', 'FO538757.3',\n",
       "       'FO538757.2', 'AP006222.2', 'RP5-857K21.15', 'RP4-669L17.2',\n",
       "       'RP4-669L17.10',\n",
       "       ...\n",
       "       'AC011043.1', 'AL592183.1', 'AC007325.1', 'AC007325.4', 'AC007325.2',\n",
       "       'BX072566.1', 'AL354822.1', 'AC023491.2', 'AC004556.1', 'AC240274.1'],\n",
       "      dtype='object', length=32146)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "75fa9076-2994-4f53-8e97-c4fb8b80b8ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>IA</th>\n",
       "      <th>TPI</th>\n",
       "      <th>GIM</th>\n",
       "      <th>SPS</th>\n",
       "      <th>AID</th>\n",
       "      <th>ERI</th>\n",
       "      <th>RCD</th>\n",
       "      <th>EGS</th>\n",
       "      <th>DCE</th>\n",
       "      <th>AIM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCR3</td>\n",
       "      <td>PLCB2</td>\n",
       "      <td>BLM</td>\n",
       "      <td>PAK2</td>\n",
       "      <td>HCST</td>\n",
       "      <td>E2F1</td>\n",
       "      <td>PAK2</td>\n",
       "      <td>PAK2</td>\n",
       "      <td>PRKAG3</td>\n",
       "      <td>CDH15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIF1A</td>\n",
       "      <td>PRKCG</td>\n",
       "      <td>FANCI</td>\n",
       "      <td>GNG3</td>\n",
       "      <td>HIF1A</td>\n",
       "      <td>PIK3CD</td>\n",
       "      <td>IL6R</td>\n",
       "      <td>GNG3</td>\n",
       "      <td>NFE2L2</td>\n",
       "      <td>PAK2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRKCG</td>\n",
       "      <td>TNFAIP3</td>\n",
       "      <td>RECQL4</td>\n",
       "      <td>IL6R</td>\n",
       "      <td>PLCB2</td>\n",
       "      <td>NRAS</td>\n",
       "      <td>PLK1</td>\n",
       "      <td>RBL1</td>\n",
       "      <td>GPI</td>\n",
       "      <td>PLCB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MMP1</td>\n",
       "      <td>CARD11</td>\n",
       "      <td>POLD3</td>\n",
       "      <td>PLCB2</td>\n",
       "      <td>CXCR1</td>\n",
       "      <td>PIK3CG</td>\n",
       "      <td>ELK1</td>\n",
       "      <td>FZD2</td>\n",
       "      <td>HIF1A</td>\n",
       "      <td>FZD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PIK3CD</td>\n",
       "      <td>GNB1</td>\n",
       "      <td>PCNA</td>\n",
       "      <td>PKN3</td>\n",
       "      <td>CCL26</td>\n",
       "      <td>PRKCA</td>\n",
       "      <td>BCL6</td>\n",
       "      <td>WNT3A</td>\n",
       "      <td>PFKP</td>\n",
       "      <td>WNT3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PIM1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CSF3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BARD1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GNA12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GNAI3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>207 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         IA      TPI     GIM    SPS    AID     ERI   RCD    EGS     DCE    AIM\n",
       "0      CCR3    PLCB2     BLM   PAK2   HCST    E2F1  PAK2   PAK2  PRKAG3  CDH15\n",
       "1     HIF1A    PRKCG   FANCI   GNG3  HIF1A  PIK3CD  IL6R   GNG3  NFE2L2   PAK2\n",
       "2     PRKCG  TNFAIP3  RECQL4   IL6R  PLCB2    NRAS  PLK1   RBL1     GPI  PLCB2\n",
       "3      MMP1   CARD11   POLD3  PLCB2  CXCR1  PIK3CG  ELK1   FZD2   HIF1A   FZD2\n",
       "4    PIK3CD     GNB1    PCNA   PKN3  CCL26   PRKCA  BCL6  WNT3A    PFKP  WNT3A\n",
       "..      ...      ...     ...    ...    ...     ...   ...    ...     ...    ...\n",
       "202     NaN      NaN     NaN   PIM1    NaN     NaN   NaN    NaN     NaN    NaN\n",
       "203     NaN      NaN     NaN   CSF3    NaN     NaN   NaN    NaN     NaN    NaN\n",
       "204     NaN      NaN     NaN  BARD1    NaN     NaN   NaN    NaN     NaN    NaN\n",
       "205     NaN      NaN     NaN  GNA12    NaN     NaN   NaN    NaN     NaN    NaN\n",
       "206     NaN      NaN     NaN  GNAI3    NaN     NaN   NaN    NaN     NaN    NaN\n",
       "\n",
       "[207 rows x 10 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_list = pd.read_csv('hallmark_surv_curated.csv')\n",
    "gene_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "76551dda-0d9e-4e37-8717-a2de00a03043",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: genes are not in var_names and ignored: ['IL8']\n",
      "WARNING: genes are not in var_names and ignored: ['GDF1']\n",
      "WARNING: genes are not in var_names and ignored: ['IL8']\n",
      "WARNING: genes are not in var_names and ignored: ['LRDD', 'TNFRSF6B']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>doublet</th>\n",
       "      <th>doublet_score</th>\n",
       "      <th>IA</th>\n",
       "      <th>TPI</th>\n",
       "      <th>GIM</th>\n",
       "      <th>SPS</th>\n",
       "      <th>AID</th>\n",
       "      <th>ERI</th>\n",
       "      <th>RCD</th>\n",
       "      <th>EGS</th>\n",
       "      <th>DCE</th>\n",
       "      <th>AIM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGGGTA-1-8</th>\n",
       "      <td>8</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.101472e-64</td>\n",
       "      <td>0.052988</td>\n",
       "      <td>0.017619</td>\n",
       "      <td>0.052130</td>\n",
       "      <td>0.114141</td>\n",
       "      <td>0.045288</td>\n",
       "      <td>0.037139</td>\n",
       "      <td>0.103837</td>\n",
       "      <td>0.081081</td>\n",
       "      <td>0.167637</td>\n",
       "      <td>0.117675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGACGCTTT-1-8</th>\n",
       "      <td>8</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.113508e+00</td>\n",
       "      <td>0.073131</td>\n",
       "      <td>0.030940</td>\n",
       "      <td>0.040435</td>\n",
       "      <td>0.109809</td>\n",
       "      <td>0.061171</td>\n",
       "      <td>0.013375</td>\n",
       "      <td>0.087562</td>\n",
       "      <td>0.069977</td>\n",
       "      <td>0.149741</td>\n",
       "      <td>0.134420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGAGCGTT-1-8</th>\n",
       "      <td>8</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.420106e-124</td>\n",
       "      <td>0.067012</td>\n",
       "      <td>0.036655</td>\n",
       "      <td>0.043667</td>\n",
       "      <td>0.102290</td>\n",
       "      <td>0.044514</td>\n",
       "      <td>0.036958</td>\n",
       "      <td>0.099073</td>\n",
       "      <td>0.080125</td>\n",
       "      <td>0.199126</td>\n",
       "      <td>0.125010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGGTTCCC-1-8</th>\n",
       "      <td>8</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.420106e-124</td>\n",
       "      <td>0.069810</td>\n",
       "      <td>0.039405</td>\n",
       "      <td>0.040295</td>\n",
       "      <td>0.111040</td>\n",
       "      <td>0.062649</td>\n",
       "      <td>0.021833</td>\n",
       "      <td>0.096528</td>\n",
       "      <td>0.082146</td>\n",
       "      <td>0.159844</td>\n",
       "      <td>0.117536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGTCGCCGT-1-8</th>\n",
       "      <td>8</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.989470e+00</td>\n",
       "      <td>0.025381</td>\n",
       "      <td>0.028524</td>\n",
       "      <td>0.038965</td>\n",
       "      <td>0.097859</td>\n",
       "      <td>0.038207</td>\n",
       "      <td>0.012125</td>\n",
       "      <td>0.087302</td>\n",
       "      <td>0.067755</td>\n",
       "      <td>0.170074</td>\n",
       "      <td>0.113355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCACAAACC-1-8</th>\n",
       "      <td>8</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.805684e-105</td>\n",
       "      <td>0.047667</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.037032</td>\n",
       "      <td>0.109142</td>\n",
       "      <td>0.026631</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.091725</td>\n",
       "      <td>0.088042</td>\n",
       "      <td>0.149948</td>\n",
       "      <td>0.107325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCATGTCTT-1-8</th>\n",
       "      <td>8</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.676102e+01</td>\n",
       "      <td>0.089512</td>\n",
       "      <td>0.052976</td>\n",
       "      <td>0.036958</td>\n",
       "      <td>0.106772</td>\n",
       "      <td>0.058486</td>\n",
       "      <td>0.048306</td>\n",
       "      <td>0.098496</td>\n",
       "      <td>0.078404</td>\n",
       "      <td>0.182881</td>\n",
       "      <td>0.130691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCGTTTATC-1-8</th>\n",
       "      <td>8</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.242906e-165</td>\n",
       "      <td>0.039976</td>\n",
       "      <td>0.060786</td>\n",
       "      <td>0.044863</td>\n",
       "      <td>0.097916</td>\n",
       "      <td>0.053757</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.080582</td>\n",
       "      <td>0.058956</td>\n",
       "      <td>0.167267</td>\n",
       "      <td>0.122906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTACCAGA-1-8</th>\n",
       "      <td>8</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.242906e-165</td>\n",
       "      <td>0.041952</td>\n",
       "      <td>0.067833</td>\n",
       "      <td>0.039660</td>\n",
       "      <td>0.110191</td>\n",
       "      <td>0.056171</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.096659</td>\n",
       "      <td>0.070630</td>\n",
       "      <td>0.153407</td>\n",
       "      <td>0.126506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTTTCCTC-1-8</th>\n",
       "      <td>8</td>\n",
       "      <td>Adipose</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.400619e-126</td>\n",
       "      <td>0.068988</td>\n",
       "      <td>0.019702</td>\n",
       "      <td>0.056961</td>\n",
       "      <td>0.100306</td>\n",
       "      <td>0.058162</td>\n",
       "      <td>0.031458</td>\n",
       "      <td>0.095287</td>\n",
       "      <td>0.067865</td>\n",
       "      <td>0.142185</td>\n",
       "      <td>0.128225</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10856 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     batch   tissue  doublet  doublet_score        IA  \\\n",
       "AAACCTGAGAAGGGTA-1-8     8  Adipose      0.0   3.101472e-64  0.052988   \n",
       "AAACCTGAGACGCTTT-1-8     8  Adipose      0.0   1.113508e+00  0.073131   \n",
       "AAACCTGAGGAGCGTT-1-8     8  Adipose      0.0  1.420106e-124  0.067012   \n",
       "AAACCTGAGGGTTCCC-1-8     8  Adipose      0.0  1.420106e-124  0.069810   \n",
       "AAACCTGAGTCGCCGT-1-8     8  Adipose      0.0   1.989470e+00  0.025381   \n",
       "...                    ...      ...      ...            ...       ...   \n",
       "TTTGTCATCACAAACC-1-8     8  Adipose      0.0  3.805684e-105  0.047667   \n",
       "TTTGTCATCATGTCTT-1-8     8  Adipose      0.0   2.676102e+01  0.089512   \n",
       "TTTGTCATCGTTTATC-1-8     8  Adipose      0.0  9.242906e-165  0.039976   \n",
       "TTTGTCATCTACCAGA-1-8     8  Adipose      0.0  9.242906e-165  0.041952   \n",
       "TTTGTCATCTTTCCTC-1-8     8  Adipose      0.0  2.400619e-126  0.068988   \n",
       "\n",
       "                           TPI       GIM       SPS       AID       ERI  \\\n",
       "AAACCTGAGAAGGGTA-1-8  0.017619  0.052130  0.114141  0.045288  0.037139   \n",
       "AAACCTGAGACGCTTT-1-8  0.030940  0.040435  0.109809  0.061171  0.013375   \n",
       "AAACCTGAGGAGCGTT-1-8  0.036655  0.043667  0.102290  0.044514  0.036958   \n",
       "AAACCTGAGGGTTCCC-1-8  0.039405  0.040295  0.111040  0.062649  0.021833   \n",
       "AAACCTGAGTCGCCGT-1-8  0.028524  0.038965  0.097859  0.038207  0.012125   \n",
       "...                        ...       ...       ...       ...       ...   \n",
       "TTTGTCATCACAAACC-1-8  0.000000  0.037032  0.109142  0.026631  0.000000   \n",
       "TTTGTCATCATGTCTT-1-8  0.052976  0.036958  0.106772  0.058486  0.048306   \n",
       "TTTGTCATCGTTTATC-1-8  0.060786  0.044863  0.097916  0.053757  0.000000   \n",
       "TTTGTCATCTACCAGA-1-8  0.067833  0.039660  0.110191  0.056171  0.000000   \n",
       "TTTGTCATCTTTCCTC-1-8  0.019702  0.056961  0.100306  0.058162  0.031458   \n",
       "\n",
       "                           RCD       EGS       DCE       AIM  \n",
       "AAACCTGAGAAGGGTA-1-8  0.103837  0.081081  0.167637  0.117675  \n",
       "AAACCTGAGACGCTTT-1-8  0.087562  0.069977  0.149741  0.134420  \n",
       "AAACCTGAGGAGCGTT-1-8  0.099073  0.080125  0.199126  0.125010  \n",
       "AAACCTGAGGGTTCCC-1-8  0.096528  0.082146  0.159844  0.117536  \n",
       "AAACCTGAGTCGCCGT-1-8  0.087302  0.067755  0.170074  0.113355  \n",
       "...                        ...       ...       ...       ...  \n",
       "TTTGTCATCACAAACC-1-8  0.091725  0.088042  0.149948  0.107325  \n",
       "TTTGTCATCATGTCTT-1-8  0.098496  0.078404  0.182881  0.130691  \n",
       "TTTGTCATCGTTTATC-1-8  0.080582  0.058956  0.167267  0.122906  \n",
       "TTTGTCATCTACCAGA-1-8  0.096659  0.070630  0.153407  0.126506  \n",
       "TTTGTCATCTTTCCTC-1-8  0.095287  0.067865  0.142185  0.128225  \n",
       "\n",
       "[10856 rows x 14 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for i in gene_list.columns:\n",
    "    genes = gene_list[i].dropna().to_list()\n",
    "    scored_adata = score_genes(adata=adata, gene_list=genes, score_name=i)\n",
    "    adata=scored_adata\n",
    "\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89818410-11ba-49db-8413-2745ab8921ed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5ff8d0b-f045-4c9c-859f-415e389bb88e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
